jupyter/sum-to-zero/sum_to_zero_evaluation.qmd

```diff
@@ -0,0 +1,1582 @@
+
+```{python}
+gen_data_dict = {
+    'N_eth':3, 'N_edu':5, 'N_age':9, 
+    'baseline': -3.5, 'sens': 0.75, 'spec': 0.9995}
+
+strata = 2 * gen_data_dict['N_age'] * gen_data_dict['N_eth'] * gen_data_dict['N_edu']
+```
+
+###### Specify the total number of observations.
+
+We generate two datasets:  one with a small number of observations, relative to the number of strata,
+and one with sufficient data to provide information on all combinations of demographics.
+
+```{python}
+gen_data = gen_data_dict.copy()
+gen_data['N'] = strata * 17
+
+gen_data_lg = gen_data_dict.copy()
+gen_data_lg['N'] = strata * 200
+
+gen_data_tiny = gen_data_dict.copy()
+gen_data_tiny['N'] = strata * 7
+```
+
+##### Run 1 sampling iteration to get a complete dataset.
+
+```{python}
+sim_data = gen_mod.sample(data=gen_data,
+                          iter_warmup=1, iter_sampling=1, chains=1, seed=45678)
+
+sim_data_lg = gen_mod.sample(data=gen_data_lg,
+                          iter_warmup=1, iter_sampling=1, chains=1, seed=45678)
+
+sim_data_tiny = gen_mod.sample(data=gen_data_tiny,
+                          iter_warmup=1, iter_sampling=1, chains=1, seed=45678)
+```
+
+##### Examine the set of generated data-generating params and resulting dataset.
+
+```{python}
+print(f'Small dataset: N = {gen_data["N"]}, strata = {strata}, expected obs per demographic {gen_data["N"] / strata}')
+print(f'Large dataset: N = {gen_data_lg["N"]}, strata = {strata}, expected obs per demographic {gen_data_lg["N"] / strata}')
+for var, value in sim_data.stan_variables().items():
+    print(var, value[0]) if isinstance(value[0], np.float64) else print(var, value[0][:10])
+```
+
+What is the distribution of the observed number of tests per demographic?
+
+```{python}
+tests = pd.Series(sim_data.stan_variable('tests')[0])
+print("Small dataset, tests per demographic", tests.describe())
+```
+
+```{python}
+tests_lg = pd.Series(sim_data_lg.stan_variable('tests')[0])
+print("Large dataset, tests per demographic", tests_lg.describe())
+```
+
+```{python}
+tests_tiny = pd.Series(sim_data_tiny.stan_variable('tests')[0])
+print("Tiny dataset, tests per demographic", tests_tiny.describe())
+```
+
+##### Plot the distribution of observed positive tests and the underlying prevalence.
+
+Because the data-generating parameters and percentage of observations per category are generated at random,
+some datasets may have very low overall disease rates and/or many unobserved strata, and will therefore be
+pathologically hard to fit.  This is informative for understanding what is consistent when
+generating a set of percentages and regression weights as is done in the Stan data generating program.
+
+```stan
+  vector[N_eth] pct_eth = dirichlet_rng(rep_vector(1, N_eth));
+  for (n in 1:N_eth) {
+    beta_eth[n] = std_normal_rng();
+  }
+```
+
+However, this can result in very unbalanced datasets, in which case it is best to
+generate another dataset and continue.
+
```

Comment:
what does the following code do vs the block after? I'm being a bit picky but it will be much clearer to the reader.

Reply:
rewrote - now stands as:

Because the data-generating parameters and percentage of observations per category are generated at random,
some datasets may have many low count or no count strata, or random effects near zero across several categories
and will therefore be pathologically hard to fit.
Using a specified seed for the data-generating program avoids this problem.

Reply:
Thanks---that looks good.